File: Design-Patterns-Using-Ruby/State/StatePattern.rb
# Define BulbState interface
module BulbState
    def flip(bc)
        raise NotImplementedError, 'Subclasses must implement the flip method'
    end

    def to_s
        raise NotImplementedError, 'Subclasses must implement the to_s method'
    end
end

# Define BulbControl class
class BulbControl
    attr_accessor :current

    def initialize
        @current = Off.new
    end

    def set_state(state)
        @current = state
    end

    def flip
        @current.flip(self)
    end

    def to_s_state
        @current.to_s
    end
end

# Define On class implementing BulbState
class On
    include BulbState

    def flip(bc)
        bc.set_state(Off.new)
    end

    def to_s
        'On'
    end
end

# Define Off class implementing BulbState
class Off
    include BulbState

    def flip(bc)
        bc.set_state(On.new)
    end

    def to_s
        'Off'
    end
end

# Client code
c = BulbControl.new
c.flip
puts c.to_s_state
c.flip
puts c.to_s_state

=begin 
On
Off
=end

File: Design-Patterns-Using-Swift/State/StatePattern.swift
protocol BulbState {
    func flip(_ bc: BulbControl)
    var description: String { get }
}

class BulbControl {
    private var current: BulbState

    init() {
        self.current = Off()
    }

    func setState(_ state: BulbState) {
        self.current = state
    }

    func flip() {
        self.current.flip(self)
    }

    var descriptionState: String {
        return self.current.description
    }
}

class On: BulbState {
    func flip(_ bc: BulbControl) {
        bc.setState(Off())
    }

    var description: String {
        return "On"
    }
}

class Off: BulbState {
    func flip(_ bc: BulbControl) {
        bc.setState(On())
    }

    var description: String {
        return "Off"
    }
}

// Client code.
let c = BulbControl()
c.flip()
print(c.descriptionState)
c.flip()
print(c.descriptionState)

/* 
On
Off
*/